### ISSUE TITLE
Passing ArrayBuffer to createWriteStream().getWriter.write() results in silent/unhelpful failure

### ISSUE THREAD

**Issue by mcclure** [jimmywarting/StreamSaver.js#181]
I am attempting to make a test app based on the https://jimmywarting.github.io/StreamSaver.js/examples/saving-a-blob.html example. I am testing in Chrome on OS X and I am serving the files off http-server on localhost. Instead of using a blob I am passing in buffers/TypedArrays directly because I happen to have them already prepared.

In my test,
1. I call `createWriteStream`, get the `getWriter()` from the result, save it as `writer`.
2. I call `await writer.write()` with an ArrayBufferView (a Uint8Array, specifically).
3. Call `await writer.write()` with an ArrayBuffer (again just because I happened to have it sitting around).
4. Call `await writer.close()`.

If I run these steps, nothing happens. Just literally nothing. I reach the end of the function (I know because I put a console.log there), but no file downloads. The network panel in the Chrome developer tools shows I downloaded several files from jimmywarting.github.io, and the console prints a weird message about MIME types (which it also prints when succeeds). But Chrome doesn't download a file or indicate a file attempted to be downloaded.

But! If I skip step 3— or even if I error out between steps 2 and 3 and skip step 4— then I download the file I expect (minus the data from step 3).  
And if instead of passing in the ArrayBuffer in in step 3, I first convert it to an ArrayBufferView by wrapping it in new Uint8Array(), then I download the file I expect (including the data from step 3. "Success".

So: My app works now. Yay! But, it is worrisome that passing an inappropriate (an ArrayBuffer) into write() results in this totally silent failure. It took me quite a while to debug this.

The error reporting here could be improved. My "expected" behavior is:
- Because passing in the wrong type is a simple/easy error to make, it would be good if it printed an error to the error console.
- One thing that would really help is if `writer.write()` had a TypeScript type more refined than "any". Demonstrably, it only accepts certain types. (I realize this might be slightly awkward to rig since `writer` is a standard WriterStreamDefaultWriter and not a type controlled by StreamSaver.)

The "error" is simple to reproduce, I could write a test case if it would be helpful.

**Comment by jimmywarting** [comment]
> I could write a test case if it would be helpful.

No need, I'm full aware of it. I wrote about arraybuffer vs uint8arrays yesterday in another issue: https://github.com/jimmywarting/StreamSaver.js/issues/180#issuecomment-699532810

But yea a `chunk instanceof Uint8array` check along with a error message could be proven useful.

> (I realize this might be slightly awkward to rig since writer is a standard WriterStreamDefaultWriter and not a type controlled by StreamSaver.)

That's why i haven't done any jsdoc or typescript for it. I wonder if there is anyway of saying *this WriterStreamDefaultWriter you get back must be of the type "byte" stream so you can only write uint8arrays* 

I'm not a fan of using typescript but i do support typing in general 
I just prefer compiler free vanilla js more. + you can accomplish most typing with just jsdoc alone

**Comment by jimmywarting** [comment]
found out something while trying out some code i VS

![Skärmavbild 2020-09-27 kl  15 45 09](https://user-images.githubusercontent.com/1148376/94366430-80df8400-00d8-11eb-9b29-c4e81cacb56c.png)

something about a underlying source<uint8array>

**Comment by mcclure** [comment]
"it all comes down to that Response constructor with a readableStream only accepts Uint8Arrays"

I see, that's helpful to know. If all else fails maybe it would be good to document that in the README.

"you can accomplish most typing with just jsdoc alone"

Is there jsdoc for the StreamSaver.js api? Is it rendered into html somewhere?

"I wonder if there is anyway of saying this WriterStreamDefaultWriter you get back must be of the type "byte" stream so you can only write uint8arrays"

I'd ask if you could subclass WriterStreamDefaultWriter somehow, but I'm a little confused how StreamSaver.js does what it is doing so I am not sure the question makes sense :)

**Comment by jimmywarting** [comment]
> Is there jsdoc for the StreamSaver.js api? Is it rendered into html somewhere?

Nope, it's not rendered into some html documentation anywhere. haven't really felt like it's worth it. The lib basically just have one function that returns a writableStream and not so much else beside a few optimal options. 
`streamSaver.createWritableStream().getWriter()`

and like you said it returns a standard WriterStreamDefaultWriter so you should already have most of the typing information from that.

> I'd ask if you could subclass WriterStreamDefaultWriter somehow, but I'm a little confused how StreamSaver.js does what it is doing so I am not sure the question makes sense :)

your question makes sense, but i don't think there is anything to subclass to fix it.

---

I just tried something and got it right on the first try:
 
![Skärmavbild 2020-09-27 kl  20 44 50](https://user-images.githubusercontent.com/1148376/94372989-805be300-0102-11eb-8cee-5e5b151da0ff.png)
![Skärmavbild 2020-09-27 kl  20 45 06](https://user-images.githubusercontent.com/1148376/94372991-818d1000-0102-11eb-8a9b-dfb5052aa15b.png)

**Comment by mcclure** [comment]
> Nope, it's not rendered into some html documentation anywhere. haven't really felt like it's worth it. The lib basically just have one function that returns a writableStream and not so much else beside a few optimal options.
streamSaver.createWritableStream({...}).getWriter() that's basically everything it dose
> 
> and like you said it returns a standard WriterStreamDefaultWriter so you should already have most of the typing information from that.

Hm, just giving some feedback as a user here, I think when I found the library I would have appreciated a proper API doc even if it was just the one function. I kind of felt like I was reverse engineering the library from the examples-- it's not obvious which options to createWriteStream() are allowed from the current doc, and it's not obvious to me that the thing it returns is a WriterStreamDefaultWriter. I had to figure those two things out from looking at the TypeScript typings, which are usually not intended for humans, and even once I figured it out it took me a bit to figure out WriterStreamDefaultWriter is a standard class with an MDN page and everything. Something shaped like an API doc would at least make it clear that the function and the 3 options are all that there is (the sample at the top of the README uses all 3 options, but it's not clear from reading the example whether there are additional options not mentioned) and would be an opportunity to link a [WritableStream](https://developer.mozilla.org/en-US/docs/Web/API/WritableStream) doc.
